src/Microsoft.Identity.Client/Cache/CacheSessionManager.cs

```diff
@@ -2,60 +2,131 @@
 // Licensed under the MIT License.
 
 using System;
+using System.Collections.Generic;
 using System.Threading.Tasks;
 using Microsoft.Identity.Client.Cache.Items;
 using Microsoft.Identity.Client.Cache.Keys;
 using Microsoft.Identity.Client.Internal.Requests;
 using Microsoft.Identity.Client.OAuth2;
+using Microsoft.Identity.Client.TelemetryCore;
+using Microsoft.Identity.Client.TelemetryCore.Internal;
+using Microsoft.Identity.Client.TelemetryCore.Internal.Events;
 
 namespace Microsoft.Identity.Client.Cache
 {
+    /// <summary>
+    /// MSAL should only interact with the cache though this object. It is reponsible for firing cache notifications.
+    /// Flows should only perform (at most) 2 cache accesses: one to read data and one to write tokens. Reading data multiple times 
+    /// (e.g. read all ATs, read all RTs) should not refresh the cache from disk because of perf impact.
+    /// Write operations are still the responsability of TokenCache (TODO)
+    /// </summary>
     internal class CacheSessionManager : ICacheSessionManager
     {
         private readonly AuthenticationRequestParameters _requestParams;
+        private readonly ITelemetryManager _telemetryManager;
+        private bool cacheRefreshedForRead = false;
 
-        public CacheSessionManager(ITokenCacheInternal tokenCacheInternal, AuthenticationRequestParameters requestParams)
+        public CacheSessionManager(
+            ITokenCacheInternal tokenCacheInternal, 
+            AuthenticationRequestParameters requestParams, 
+            ITelemetryManager telemetryManager)
         {
             TokenCacheInternal = tokenCacheInternal;
-            _requestParams = requestParams;
+            _requestParams = requestParams ?? throw new ArgumentNullException(nameof(requestParams));
+            _telemetryManager = telemetryManager ?? throw new ArgumentNullException(nameof(telemetryManager));
         }
 
+        #region ICacheSessionManager implementation
         public ITokenCacheInternal TokenCacheInternal { get; }
-        public bool HasCache => TokenCacheInternal != null;
 
-        public Task<MsalAccessTokenCacheItem> FindAccessTokenAsync()
+        public bool HasCache => TokenCacheInternal != null; 
+
+        public async Task<MsalAccessTokenCacheItem> FindAccessTokenAsync()
```

Comment:
MSAL requests will interact with the cache only via the CacheSessionManager. Today, this is messy - some logic uses CacheSessionManager, some use the Accessor